### PR TITLE
Improve the performance of tiled decompression

### PR DESCRIPTION
**Pull request by Cadair**
This is a work in progress for the moment as we continue to iterate on the performance hot spots.

### CODE REVIEW

**Comment by astrofrog** [comment]
We should try and re-write _iter_array_tiles to not use Numpy as I think this slows things down a bit

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 1]
```diff
@@ -269,79 +310,121 @@ def decompress_hdu_section(hdu, first_tile_index, last_tile_index):
...
+        heap_cache = hdu._get_raw_data(hdu._header["PCOUNT"], np.uint8, hdu._data_offset + hdu._theap)
+    else:
+        heap_cache = None
+
+        # if size == 0:
+        #     return np.array([], dtype=dtype)
+        # raw = self._get_data_from_heap(offset, size, dtype)
+        # Return tile in native endian since this is what is expected
+        # by the decompression functions
+        # return raw.astype(raw.dtype.newbyteorder("="), copy=False)
```
Comment (astrofrog) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:354]: Oops need to remove this

File: astropy/io/fits/_tiled_compression/codecs.py [hunk 2]
```diff
@@ -34,9 +34,64 @@ class Codec:
     "Rice1",
     "PLIO1",
     "HCompress1",
+    "NoCompress",
 ]
 
 
+def _as_big_endian_array(data):
+    return data.astype(np.asarray(data).dtype.newbyteorder(">"))
```
Comment (Cadair) [inline astropy/io/fits/_tiled_compression/codecs.py:42]: @astrofrog isn't this going to copy the array?

Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/codecs.py:42]: Good catch need copy=False

File: astropy/io/fits/_tiled_compression/tiled_compression.py (hunk 1 above)
Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/tiled_compression.py]: Done

File: astropy/io/fits/_tiled_compression/codecs.py (hunk 2 above)
Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/codecs.py]: Fixed (was done in the NOCOMPRESS PR)

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 3]
```diff
@@ -263,86 +302,137 @@ def decompress_hdu_section(hdu, first_tile_index, last_tile_index):
...
-    data = np.zeros(buffer_shape, dtype=BITPIX2DTYPE[hdu._header["ZBITPIX"]])
+    data = np.empty(buffer_shape, dtype=BITPIX2DTYPE[hdu._header["ZBITPIX"]])
 
     quantized = "ZSCALE" in hdu.compressed_data.dtype.names
 
     if data.size == 0:
         return data
 
+    settings = _header_to_settings(hdu._header)
+    ZCMPTYPE = hdu._header["ZCMPTYPE"]
```
Comment (astrofrog) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:313]: In decompress, you called this ``comp_type`` - maybe we should make consistent?

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 4]
```diff
@@ -263,86 +302,137 @@ def decompress_hdu_section(hdu, first_tile_index, last_tile_index):
...
+        zzero_column = None
+
+    zblank_header = hdu._header.get("ZBLANK", None)
+
+    gzip_compressed_data_column = None
+    gzip_compressed_data_dtype = None
+
+    # If more than half the data is requested, read in all the heap.
+    # TODO: decide what heuristic we actually want here
+    if np.product(buffer_shape) > 0.5 * np.product(data_shape):
```
Comment (astrofrog) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:346]: @Cadair - we need to think about this!

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 5]
```diff
@@ -126,8 +140,10 @@ def _finalize_array(tile_buffer, *, bitpix, tile_shape, algorithm, lossless):
     else:
         # For RICE_1 compression the tiles that are on the edge can end up
         # being padded, so we truncate excess values
-        if algorithm in ("RICE_1", "RICE_ONE", "PLIO_1"):
-            tile_buffer = tile_buffer[: np.product(tile_shape)]
+        if algorithm in ("RICE_1", "RICE_ONE", "PLIO_1") and tile_size < len(
+            tile_buffer
+        ):
```
Comment (Cadair) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:145]: :nauseated_face:

**Comment by astrofrog** [comment]
@saimn - I believe this is ready for review - this speeds things up by a factor of a few in our new tiled compression algorithm, to bring it a lot closer to the original performance in previous astropy versions.

File: astropy/io/fits/_tiled_compression/codecs.py [hunk 6]
```diff
@@ -334,7 +334,7 @@ def decode(self, buf):
         buf : np.ndarray
             The decompressed buffer.
         """
-        cbytes = np.frombuffer(buf, dtype=np.uint8).tobytes()
+        cbytes = np.frombuffer(_as_native_endian_array(buf), dtype=np.uint8).tobytes()
```
Comment (saimn) [inline astropy/io/fits/_tiled_compression/codecs.py:337]: Why only for this one and not for RICE or HCOMPRESS ?

File: astropy/io/fits/_tiled_compression/tiled_compression.py [hunk 7]
```diff
@@ -65,23 +65,36 @@ def _compress_tile(buf, *, algorithm: str, **settings):
...
+
+def _update_tile_settings(settings, compression_type, actual_tile_shape):
+    """
+    Update the settings with tile-specific settings
+    """
+    if compression_type == "PLIO_1":
+        # We have to calculate the tilesize from the shape of the tile not the
+        # header, so that it's correct for edge tiles etc.
+        settings["tilesize"] = prod(actual_tile_shape)
+    elif compression_type in ("RICE_1", "RICE_ONE"):
```
Comment (saimn) [inline astropy/io/fits/_tiled_compression/tiled_compression.py:95]: Merge PLIO_1 with this one ?

Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/tiled_compression.py]: Done!

File: astropy/io/fits/_tiled_compression/codecs.py (hunk 6 above)
Comment (astrofrog) [reply astropy/io/fits/_tiled_compression/codecs.py:337]: Good catch, thanks!

**Comment by astrofrog** [comment]
@saimn - I've addressed your comments 😄

**Review by saimn — APPROVED** [review]
Nice improvements, thanks @Cadair & @astrofrog !

**Pull request merged**